src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,510 @@
+                _sessionToken = sessionToken;
+                _randomByteGenerator = Ensure.IsNotNull(randomByteGenerator, nameof(randomByteGenerator));
+                _clock = Ensure.IsNotNull(clock, nameof(clock));
+            }
+
+            public string Name
+            {
+                get { return MechanismName; }
+            }
+
+            public ISaslStep Initialize(IConnection connection, SaslConversation conversation, ConnectionDescription description)
+            {
+                Ensure.IsNotNull(connection, nameof(connection));
+                Ensure.IsNotNull(description, nameof(description));
+
+                var nonce = GenerateRandomBytes();
+
+                var document = new BsonDocument
+                {
+                    { "r", nonce },
+                    { "p", (int)'n' }
+                };
+
+                var clientMessageBytes = document.ToBson();
+
+                return new ClientFirst(clientMessageBytes, nonce, _credential, _sessionToken, _clock);
+            }
+
+            private byte[] GenerateRandomBytes()
+            {
+                return _randomByteGenerator.Generate(ClientNonceLength);
+            }
+        }
+
+        private class ClientFirst : ISaslStep
+        {
+            private readonly byte[] _bytesToSendToServer;
+            private readonly IClock _clock;
+            private readonly UsernamePasswordCredential _credential;
+            private readonly byte[] _nonce;
+            private readonly string _sessionToken;
+
+            public ClientFirst(
+                byte[] bytesToSendToServer,
+                byte[] nonce,
+                UsernamePasswordCredential credential,
+                string sessionToken,
+                IClock clock)
+            {
+                _bytesToSendToServer = bytesToSendToServer;
+                _clock = clock;
+                _credential = credential;
+                _nonce = nonce;
+                _sessionToken = sessionToken;
+            }
+
+            public byte[] BytesToSendToServer
+            {
+                get { return _bytesToSendToServer; }
+            }
+
+            public bool IsComplete
+            {
+                get { return false; }
+            }
+
+            public ISaslStep Transition(SaslConversation conversation, byte[] bytesReceivedFromServer)
+            {
+                var serverFirstMessageDocument = BsonSerializer.Deserialize<BsonDocument>(bytesReceivedFromServer);
+                var serverNonce = serverFirstMessageDocument["s"].AsByteArray;
+                var host = serverFirstMessageDocument["h"].AsString;
+
+                if (serverNonce.Length != ClientNonceLength * 2 || !serverNonce.Take(ClientNonceLength).SequenceEqual(_nonce))
+                {
+                    throw new MongoAuthenticationException(conversation.ConnectionId, message: "Server sent an invalid nonce.");
```

Comment:
`message:` is not really necessary.

It's hard to say when these optional parameter names are helpful or not.